README.md
<div align="center">
  <h1>JavaScript Examples</h1>
  Collection of examples in <a href="https://en.wikipedia.org/wiki/JavaScript">JavaScript</a> to show how to get things done in an idiomatic way
</div>

## Installing

You need to have [Node.js] **v10.0.0** or higher installed to run the examples.

Some examples depend on [npm] packages so, make sure you install everything before trying to run the examples.
You can do this with the following commands.

```
$ git clone https://github.com/umut-sahin/javascript-examples.git
$ cd javascript-examples
$ npm install
```

## Running a specific example

```
$ npm run -s <example-name>
```

Also, you can pass arguments like so

```
$ npm run -s <example-name> -- [arguments...]
```

## Available examples

- [brainfuck-interpreter]
- [command-line-arguments]
- [http-hello-world-server]
- [read-from-console]
- [tcp-echo-server]
- [timers]
- [write-to-console]

## Changelog

See [CHANGELOG].

## License

This repository is released under [MIT license].
Please see [LICENSE] for details.

## Contributing

Contributions are welcome!
Please see [CONTRIBUTING] for guidance.

## Authors

See [AUTHORS].


[//]: # (Links)

[AUTHORS]:
  https://github.com/umut-sahin/javascript-examples/blob/master/AUTHORS.md
[brainfuck-interpreter]:
  https://github.com/umut-sahin/javascript-examples/tree/master/brainfuck-interpreter
[CHANGELOG]:
  https://github.com/umut-sahin/javascript-examples/blob/master/CHANGELOG.md
[command-line-arguments]:
  https://github.com/umut-sahin/javascript-examples/tree/master/command-line-arguments
[CONTRIBUTING]:
  https://github.com/umut-sahin/javascript-examples/blob/master/CONTRIBUTING.md
[http-hello-world-server]:
  https://github.com/umut-sahin/javascript-examples/tree/master/http-hello-world-server
[LICENSE]:
  https://github.com/umut-sahin/javascript-examples/blob/master/LICENSE
[MIT License]:
  https://tldrlegal.com/license/mit-license
[Node.js]:
  https://nodejs.org/
[npm]:
  https://www.npmjs.com/
[read-from-console]:
  https://github.com/umut-sahin/javascript-examples/tree/master/read-from-console
[tcp-echo-server]:
  https://github.com/umut-sahin/javascript-examples/tree/master/tcp-echo-server
[timers]:
  https://github.com/umut-sahin/javascript-examples/tree/master/timers
[write-to-console]:
  https://github.com/umut-sahin/javascript-examples/tree/master/write-to-console


CHANGELOG.md
1.0.1 / 16-08-2019
==================

  - Various typos in READMEs are fixed. ([#2](https://github.com/umut-sahin/javascript-examples/issues/2))

1.0.0 / 15-07-2019
==================

  - The project is moved to [GitHub].
  - [brainfuck-interpreter] example is added.
  - [command-line-arguments] example is added.
  - [http-hello-world-server] example is added.
  - [read-from-console] example is added.
  - [tcp-echo-server] example is added.
  - [timers] example is added.
  - [write-to-console] example is added.


[//]: # (Links)

[brainfuck-interpreter]:
  https://github.com/umut-sahin/javascript-examples/tree/master/brainfuck-interpreter
[command-line-arguments]:
  https://github.com/umut-sahin/javascript-examples/tree/master/command-line-arguments
[GitHub]:
  https://github.com
[http-hello-world-server]:
  https://github.com/umut-sahin/javascript-examples/tree/master/http-hello-world-server
[read-from-console]:
  https://github.com/umut-sahin/javascript-examples/tree/master/read-from-console
[tcp-echo-server]:
  https://github.com/umut-sahin/javascript-examples/tree/master/tcp-echo-server
[timers]:
  https://github.com/umut-sahin/javascript-examples/tree/master/timers
[write-to-console]:
  https://github.com/umut-sahin/javascript-examples/tree/master/write-to-console


brainfuck-interpreter/test/instruction.js
import test from "ava";

import Instruction from "../src/instruction";


// region static get Instruction.type

test("static get Instruction.type", (t) => {
  t.deepEqual(
    Object.keys(Instruction.type),
    [
      "MOVE_LEFT",
      "MOVE_RIGHT",
      "INCREMENT",
      "DECREMENT",
      "READ",
      "WRITE",
      "START_LOOP",
      "END_LOOP",
    ],
  );
});

// endregion


// region Instruction.toString(...)

function toString(t, inputs, output) {
  t.is(new Instruction(inputs[0], inputs[1] || 0).toString(), output);
}
toString.title = (subtitle = "", inputs) => {
  const instructionName =
    Object.keys(Instruction.type).find(key => Instruction.type[key] === inputs[0]);
  const maybeData = inputs[1] ? `, data: ${inputs[1]}` : "";
  return `${subtitle} Instruction.toString() @ { type: ${instructionName}${maybeData} }`.trim();
};

test(toString, [Instruction.type.MOVE_LEFT, 0], "");
test(toString, [Instruction.type.MOVE_LEFT, 1], "<");
test(toString, [Instruction.type.MOVE_LEFT, 3], "<<<");

test(toString, [Instruction.type.MOVE_RIGHT, 0], "");
test(toString, [Instruction.type.MOVE_RIGHT, 1], ">");
test(toString, [Instruction.type.MOVE_RIGHT, 3], ">>>");

test(toString, [Instruction.type.INCREMENT, 0], "");
test(toString, [Instruction.type.INCREMENT, 1], "+");
test(toString, [Instruction.type.INCREMENT, 3], "+++");

test(toString, [Instruction.type.DECREMENT, 0], "");
test(toString, [Instruction.type.DECREMENT, 1], "-");
test(toString, [Instruction.type.DECREMENT, 3], "---");

test(toString, [Instruction.type.READ], ",");

test(toString, [Instruction.type.WRITE], ".");

test(toString, [Instruction.type.START_LOOP, 0], "[");
test(toString, [Instruction.type.START_LOOP, 1], "[");
test(toString, [Instruction.type.START_LOOP, 3], "[");

test(toString, [Instruction.type.END_LOOP, 0], "]");
test(toString, [Instruction.type.END_LOOP, 1], "]");
test(toString, [Instruction.type.END_LOOP, 3], "]");

// endregion


brainfuck-interpreter/test/interpreter.js
import test from "ava";

import Interpreter, { __private__ } from "../src/interpreter";


// region private function wrappingAddOnByte(...)

function wrappingAddOnByte(t, inputs, output) {
  t.is(__private__.wrappingAddOnByte(inputs[0], inputs[1]), output);
}
wrappingAddOnByte.title =
  (subtitle = "", inputs) =>
    `${subtitle} private function wrappingAddOnByte(${inputs[0]}, ${inputs[1]})`.trim();

test(wrappingAddOnByte, [0, 0], 0);
test(wrappingAddOnByte, [0, 1], 1);
test(wrappingAddOnByte, [0, 2], 2);
test(wrappingAddOnByte, [0, 253], 253);
test(wrappingAddOnByte, [0, 254], 254);
test(wrappingAddOnByte, [0, 255], 255);

test(wrappingAddOnByte, [1, 0], 1);
test(wrappingAddOnByte, [1, 1], 2);
test(wrappingAddOnByte, [1, 2], 3);
test(wrappingAddOnByte, [1, 253], 254);
test(wrappingAddOnByte, [1, 254], 255);
test(wrappingAddOnByte, [1, 255], 0);

test(wrappingAddOnByte, [2, 0], 2);
test(wrappingAddOnByte, [2, 1], 3);
test(wrappingAddOnByte, [2, 2], 4);
test(wrappingAddOnByte, [2, 253], 255);
test(wrappingAddOnByte, [2, 254], 0);
test(wrappingAddOnByte, [2, 255], 1);

test(wrappingAddOnByte, [253, 0], 253);
test(wrappingAddOnByte, [253, 1], 254);
test(wrappingAddOnByte, [253, 2], 255);
test(wrappingAddOnByte, [253, 253], 250);
test(wrappingAddOnByte, [253, 254], 251);
test(wrappingAddOnByte, [253, 255], 252);

test(wrappingAddOnByte, [254, 0], 254);
test(wrappingAddOnByte, [254, 1], 255);
test(wrappingAddOnByte, [254, 2], 0);
test(wrappingAddOnByte, [254, 253], 251);
test(wrappingAddOnByte, [254, 254], 252);
test(wrappingAddOnByte, [254, 255], 253);

test(wrappingAddOnByte, [255, 0], 255);
test(wrappingAddOnByte, [255, 1], 0);
test(wrappingAddOnByte, [255, 2], 1);
test(wrappingAddOnByte, [255, 253], 252);
test(wrappingAddOnByte, [255, 254], 253);
test(wrappingAddOnByte, [255, 255], 254);

// endregion

// region private function wrappingSubOnByte(...)

function wrappingSubOnByte(t, inputs, output) {
  t.is(__private__.wrappingSubOnByte(inputs[0], inputs[1]), output);
}
wrappingSubOnByte.title =
  (subtitle = "", inputs) =>
    `${subtitle} private function wrappingSubOnByte(${inputs[0]}, ${inputs[1]})`.trim();

test(wrappingSubOnByte, [0, 0], 0);
test(wrappingSubOnByte, [0, 1], 255);
test(wrappingSubOnByte, [0, 2], 254);
test(wrappingSubOnByte, [0, 253], 3);
test(wrappingSubOnByte, [0, 254], 2);
test(wrappingSubOnByte, [0, 255], 1);

test(wrappingSubOnByte, [1, 0], 1);
test(wrappingSubOnByte, [1, 1], 0);
test(wrappingSubOnByte, [1, 2], 255);
test(wrappingSubOnByte, [1, 253], 4);
test(wrappingSubOnByte, [1, 254], 3);
test(wrappingSubOnByte, [1, 255], 2);

test(wrappingSubOnByte, [2, 0], 2);
test(wrappingSubOnByte, [2, 1], 1);
test(wrappingSubOnByte, [2, 2], 0);
test(wrappingSubOnByte, [2, 253], 5);
test(wrappingSubOnByte, [2, 254], 4);
test(wrappingSubOnByte, [2, 255], 3);

test(wrappingSubOnByte, [253, 0], 253);
test(wrappingSubOnByte, [253, 1], 252);
test(wrappingSubOnByte, [253, 2], 251);
test(wrappingSubOnByte, [253, 253], 0);
test(wrappingSubOnByte, [253, 254], 255);
test(wrappingSubOnByte, [253, 255], 254);

test(wrappingSubOnByte, [254, 0], 254);
test(wrappingSubOnByte, [254, 1], 253);
test(wrappingSubOnByte, [254, 2], 252);
test(wrappingSubOnByte, [254, 253], 1);
test(wrappingSubOnByte, [254, 254], 0);
test(wrappingSubOnByte, [254, 255], 255);

test(wrappingSubOnByte, [255, 0], 255);
test(wrappingSubOnByte, [255, 1], 254);
test(wrappingSubOnByte, [255, 2], 253);
test(wrappingSubOnByte, [255, 253], 2);
test(wrappingSubOnByte, [255, 254], 1);
test(wrappingSubOnByte, [255, 255], 0);

// endregion


// region static get Interpreter.MEMORY_SIZE

test("static get Interpreter.MEMORY_SIZE", (t) => {
  t.is(Interpreter.MEMORY_SIZE, 30000);
});

// endregion


brainfuck-interpreter/src/instruction.js
import assert from "assert";

export default class Instruction {
  static get type() {
    return {
      MOVE_LEFT: 1,
      MOVE_RIGHT: 2,
      INCREMENT: 3,
      DECREMENT: 4,
      READ: 5,
      WRITE: 6,
      START_LOOP: 7,
      END_LOOP: 8,
    };
  }


  constructor(type, data = 0) {
    assert(Object.values(Instruction.type).includes(type));

    this._type = type;
    this.data = data;
  }


  toString() {
    switch (this._type) {
      case Instruction.type.MOVE_LEFT: {
        return "<".repeat(this.data);
      }
      case Instruction.type.MOVE_RIGHT: {
        return ">".repeat(this.data);
      }
      case Instruction.type.INCREMENT: {
        return "+".repeat(this.data);
      }
      case Instruction.type.DECREMENT: {
        return "-".repeat(this.data);
      }
      case Instruction.type.READ: {
        return ",";
      }
      case Instruction.type.WRITE: {
        return ".";
      }
      case Instruction.type.START_LOOP: {
        return "[";
      }
      case Instruction.type.END_LOOP: {
        return "]";
      }
      default: {
        assert(false);
        return "";
      }
    }
  }
}


brainfuck-interpreter/src/interpreter.js
import * as colors from "ansi-colors";
import fs from "fs";

import Instruction from "./instruction.js";


function wrappingAddOnByte(number, amount) {
  const available = 255 - number;

  if (amount > available) {
    return amount - (available + 1);
  }

  return number + amount;
}

function wrappingSubOnByte(number, amount) {
  if (amount > number) {
    return 255 - (amount - number - 1);
  }

  return number - amount;
}

export const __private__ = { wrappingAddOnByte, wrappingSubOnByte };


export default class Interpreter {
  static get MEMORY_SIZE() {
    return 30000;
  }


  constructor(script = "") {
    this.program = [];
    this.load(script);
  }


  execute() {
    let stdin;
    try {
      stdin = fs.openSync("/dev/stdin", "rs");
    } catch (error) {
      throw new Error(
        `${colors.red("RuntimeError:")} Unexpected IO error occurred (${error.message})`,
      );
    }

    const readWriteBuffer = Buffer.alloc(1);

    const tape = new Array(Interpreter.MEMORY_SIZE).fill(0);
    let readingHeadLocation = 0;

    let currentInstructionIndex = 0;
    while (currentInstructionIndex !== this.program.length) {
      const currentInstruction = this.program[currentInstructionIndex];
      switch (currentInstruction._type) {
        case Instruction.type.MOVE_LEFT: {
          if (currentInstruction.data > readingHeadLocation) {
            fs.closeSync(stdin);
            throw new Error(`${colors.red("RuntimeError:")} Attempted to access a negative cell`);
          }
          readingHeadLocation -= currentInstruction.data;
          break;
        }
        case Instruction.type.MOVE_RIGHT: {
          if (currentInstruction.data > Interpreter.MEMORY_SIZE - (readingHeadLocation + 1)) {
            fs.closeSync(stdin);
            throw new Error(
              `${colors.red("RuntimeError:")} Attempted to access a cell, which is above the cell limit`,
            );
          }
          readingHeadLocation += currentInstruction.data;
          break;
        }
        case Instruction.type.INCREMENT: {
          tape[readingHeadLocation] =
            wrappingAddOnByte(tape[readingHeadLocation], currentInstruction.data);
          break;
        }
        case Instruction.type.DECREMENT: {
          tape[readingHeadLocation] =
            wrappingSubOnByte(tape[readingHeadLocation], currentInstruction.data);
          break;
        }
        case Instruction.type.READ: {
          try {
            if (fs.readSync(stdin, readWriteBuffer, 0, 1, null) === 0) {
              tape[readingHeadLocation] = 0;
            } else {
              tape[readingHeadLocation] = readWriteBuffer[0];
            }
          } catch (error) {
            fs.closeSync(stdin);
            throw new Error(
              `${colors.red("RuntimeError:")} Unexpected IO error occurred (${error.message})`,
            );
          }
          break;
        }
        case Instruction.type.WRITE: {
          try {
            readWriteBuffer[0] = tape[readingHeadLocation];
            process.stdout.write(readWriteBuffer);
          } catch (error) {
            fs.closeSync(stdin);
            throw new Error(
              `${colors.red("RuntimeError:")} Unexpected IO error occurred (${error.message})`,
            );
          }
          break;
        }
        case Instruction.type.START_LOOP: {
          if (tape[readingHeadLocation] === 0) {
            currentInstructionIndex = currentInstruction.data;
          }
          break;
        }
        case Instruction.type.END_LOOP: {
          if (tape[readingHeadLocation] !== 0) {
            currentInstructionIndex = currentInstruction.data;
          }
          break;
        }
      }
      currentInstructionIndex += 1;
    }

    fs.closeSync(stdin);
    return tape;
  }

  load(script) {
    let currentLine = 1;
    let currentColumn = 1;

    const loopBalancer = [];

    const newProgram = [];
    for (const token of script) {
      switch (token) {
        case "<": {
          if (newProgram.length === 0) {
            newProgram.push(new Instruction(Instruction.type.MOVE_LEFT, 1));
            continue;
          }
          const previousInstruction = newProgram[newProgram.length - 1];
          switch (previousInstruction._type) {
            case Instruction.type.MOVE_LEFT: {
              previousInstruction.data++;
              break;
            }
            case Instruction.type.MOVE_RIGHT: {
              previousInstruction.data--;
              if (previousInstruction.data === 0) {
                newProgram.pop();
              }
              break;
            }
            default: {
              newProgram.push(new Instruction(Instruction.type.MOVE_LEFT, 1));
              break;
            }
          }
          break;
        }
        case ">": {
          if (newProgram.length === 0) {
            newProgram.push(new Instruction(Instruction.type.MOVE_RIGHT, 1));
            continue;
          }
          const previousInstruction = newProgram[newProgram.length - 1];
          switch (previousInstruction._type) {
            case Instruction.type.MOVE_RIGHT: {
              previousInstruction.data++;
              break;
            }
            case Instruction.type.MOVE_LEFT: {
              previousInstruction.data--;
              if (previousInstruction.data === 0) {
                newProgram.pop();
              }
              break;
            }
            default: {
              newProgram.push(new Instruction(Instruction.type.MOVE_RIGHT, 1));
              break;
            }
          }
          break;
        }
        case "+": {
          if (newProgram.length === 0) {
            newProgram.push(new Instruction(Instruction.type.INCREMENT, 1));
            continue;
          }
          const previousInstruction = newProgram[newProgram.length - 1];
          switch (previousInstruction._type) {
            case Instruction.type.INCREMENT: {
              previousInstruction.data = wrappingAddOnByte(previousInstruction.data, 1);
              if (previousInstruction.data === 0) {
                newProgram.pop();
              }
              break;
            }
            case Instruction.type.DECREMENT: {
              previousInstruction.data--;
              if (previousInstruction.data === 0) {
                newProgram.pop();
              }
              break;
            }
            default: {
              newProgram.push(new Instruction(Instruction.type.INCREMENT, 1));
              break;
            }
          }
          break;
        }
        case "-": {
          if (newProgram.length === 0) {
            newProgram.push(new Instruction(Instruction.type.DECREMENT, 1));
            continue;
          }
          const previousInstruction = newProgram[newProgram.length - 1];
          switch (previousInstruction.type) {
            case Instruction.type.DECREMENT: {
              previousInstruction.data = wrappingAddOnByte(previousInstruction.data, 1);
              if (previousInstruction.data === 0) {
                newProgram.pop();
              }
              break;
            }
            case Instruction.type.INCREMENT: {
              previousInstruction.data--;
              if (previousInstruction.data === 0) {
                newProgram.pop();
              }
              break;
            }
            default: {
              newProgram.push(new Instruction(Instruction.type.DECREMENT, 1));
              break;
            }
          }
          break;
        }
        case ",": {
          newProgram.push(new Instruction(Instruction.type.READ));
          break;
        }
        case ".": {
          newProgram.push(new Instruction(Instruction.type.WRITE));
          break;
        }
        case "[": {
          loopBalancer.push([newProgram.length, currentLine, currentColumn]);
          newProgram.push(new Instruction(Instruction.type.START_LOOP, 0));
          break;
        }
        case "]": {
          if (loopBalancer.length === 0) {
            throw new Error(
              `${colors.red("SyntaxError:")} Unable to find the opening bracket of ']' at ${currentLine}:${currentColumn}`,
            );
          }
          const [matchingStartLoopInstructionIndex] = loopBalancer.pop();
          newProgram[matchingStartLoopInstructionIndex].data = newProgram.length;
          newProgram.push(
            new Instruction(Instruction.type.END_LOOP, matchingStartLoopInstructionIndex),
          );
          break;
        }
        case "\n": {
          currentLine++;
          currentColumn = 0;
          break;
        }
        default: {
          break;
        }
      }
      currentColumn += 1;
    }

    if (loopBalancer.length !== 0) {
      const [, line, column] = loopBalancer.pop();
      throw new Error(
        `${colors.red("SyntaxError:")} Unable to find the closing bracket of '[' at ${line}:${column}`,
      );
    }

    this.program = newProgram;
  }

  toString() {
    return this.program.map(Instruction.toString).join();
  }
}
